kv/indexer.go

```diff
@@ -13,71 +13,84 @@ type kvIndexer struct {
 	ctx       context.Context
 	cancel    context.CancelFunc
 	indexChan chan indexBatch
-	finished  chan struct{}
-	oncer     sync.Once
+	wg        sync.WaitGroup
+	working   chan struct{}
 }
 
 type indexBatch struct {
 	bucketName []byte
-	keys       [][]byte
+	idxs       map[string][]byte
 }
 
 func NewIndexer(log *zap.Logger, kv Store) *kvIndexer {
 	ctx, cancel := context.WithCancel(context.Background())
-	i := &kvIndexer{
+	return &kvIndexer{
 		log:       log,
 		kv:        kv,
 		ctx:       ctx,
 		cancel:    cancel,
 		indexChan: make(chan indexBatch, 10),
-		finished:  make(chan struct{}),
+		working:   make(chan struct{}, 1),
 	}
-
-	go i.workIndexes()
-	return i
 }
 
-func (i *kvIndexer) AddToIndex(bucketName []byte, keys [][]byte) {
+func (i *kvIndexer) AddToIndex(bucketName []byte, idxs map[string][]byte) {
 	// check for close
 	select {
 	case <-i.ctx.Done():
 		return
-	case i.indexChan <- indexBatch{bucketName, keys}:
+	case i.indexChan <- indexBatch{bucketName, idxs}:
+	}
+
+	// add to the waitgroup and start the work process
+	select {
+	case i.working <- struct{}{}:
+		// i was able to insert i should start a worker
+		i.wg.Add(1)
+		go i.workIndexes()
+	default:
+		// we have reached our worker limit and we cannot start any more.
+		return
 	}
 }
 
 func (i *kvIndexer) workIndexes() {
-	defer close(i.finished)
-	for batch := range i.indexChan {
-		// open update tx
-		err := i.kv.Update(i.ctx, func(tx Tx) error {
-			// create a bucket for this batch
-			bucket, err := tx.Bucket(batch.bucketName)
-			if err != nil {
-				return err
-			}
-			// insert all the keys
-			for _, key := range batch.keys {
-				err := bucket.Put(key, nil)
+	// let the system know we have finished
+	defer i.wg.Done()
+	// releasee the worker hold so the system can start more later
+	defer func() { <-i.working }()
+
+	for {
+		select {
+		case batch := <-i.indexChan:
+			// open update tx
+			err := i.kv.Update(i.ctx, func(tx Tx) error {
+				// create a bucket for this batch
+				bucket, err := tx.Bucket(batch.bucketName)
 				if err != nil {
 					return err
 				}
-			}
-			return nil
-		})
+				// insert all the keys
+				for k, v := range batch.idxs {
+					err := bucket.Put([]byte(k), v)
```

Comment:
How will we deal with indexing 1 to many here? I imagine that we would really only be interested in creating a composite key here of the thing to index on (foreign) combined with the target indexed item (primary) via some separator. I did this in the generic on here: https://gist.github.com/GeorgeMac/143bd3d1ecab194bada9ff01341e9907#file-indexer-go-L61-L64

This went through my mind when I did the inline one and not when I implemented it for auth lookup for some reason 😂 . Do we want to have the abstraction take a separate `foreign -> primary` mapping and apply a standard here? Or have the caller make a composite key and just send it here to be inserted?
My gut is to have the Indexer take the two separately and combine them itself.

Reply:
Ahh OK I read further down. You're building the composite from the outside cool 👍